### ISSUE TITLE
Switch e2e matmul tool to using explicit transfer operations instead of mapping.

### ISSUE THREAD

**Issue by benvanik** [google/iree#8784]
`iree-e2e-matmul-test.cc` currently requires that all buffers are mappable into host memory in its routines to read, write, and compare memory. This worked on the local CPU path because all memory was local but in the recently added GPU path it was relying on the compiler and CUDA HAL using concurrently accessible shared memory. This was always sketchy but would work so long as the CUDA device had the `CU_DEVICE_ATTRIBUTE_CONCURRENT_MANAGED_ACCESS` cap - not all do (mine doesn't). Managed memory - especially concurrently accessible managed memory - is a big performance foot-gun and I'm working to remove our use of it - we originally had to support it because all of the tools mapped memory but now they've been updated and the matmul test is the last remaining user requiring it.

Instead of mapping the buffers for read/write/compare the tool should use one of the `iree_hal_device_transfer_*` methods:
https://github.com/google/iree/blob/c1d6af3f2b73de00e004a7d47df98ee23d8d9705/iree/hal/device.h#L211-L278
These allow for easy synchronous host<->device/device<->device transfers that are fine to use in tooling like this.

For initial buffer population the most efficient/easiest approach is to use `iree_hal_buffer_view_generate_buffer`:
https://github.com/google/iree/blob/77d32366715396797cc673653fe23d7ddabb899b/iree/hal/buffer_view_util.h#L71-L99

This will avoid extra allocations/copies if mapping is possible and otherwise setup the staging memory and perform the transfer as needed.

Basically, iree-e2e-matmul-test should not assume that any `iree_hal_buffer_map_*` methods will work. If a real application concerned with performance it could conditionally check to see if the buffer can be mapped and then take that path, but given that the operations are not performance sensitive and the sizes being copied are so small it'll be simpler to just use the utility methods.

**Comment by bjacob** [comment]
If the IREE API allowed to write test that had a portability problem, it probably also allows for end-users' code to have portability problems. Maybe enabling non-universally-supported IREE_HAL_BUFFER_USAGE_ bits should be guarded by some explicit enabling mechanism. Maybe that could be connected to runtime device detection in some way.
